### PR TITLE
[StructuralMechanics] Shell Base Element

### LINKED ISSUE #1509: Repeated defines in shell elements
```C++
#define OPT_NUM_NODES
#define OPT_STRAIN_SIZE
#define OPT_NUM_DOFS
```

These defines are multiple times defined across different shell elements in StructuralMechanics application.

A possible solution for this would be to use a `static constexpr` inside the header of the element to store the values and then using it from its class:

Header file:
```C++
class ShellThickElement3D3N : public Element {
    static constexpr int OPT_NUM_NODES = 3
    static constexpr int OPT_STRAIN_SIZE = 6
    static constexpr int OPT_NUM_DOFS = 18
    ... // rest of the class
}
```

Source file:
```C++
// This:
if (geom.PointsNumber() != OPT_NUM_NODES)

// Turns into this:
if (geom.PointsNumber() != ShellThickElement3D3N::OPT_NUM_NODES)
```

### PR DESCRIPTION
**Pull request by philbucher**
This PR introduced a BaseClass for the thin and thick shells in order to reduce the code duplication that was/is very extensive in the shells.

A lot of functions (mostly the public ones) were moved to the BaseClass, I did not change any implementational details.

In a next PR I plan to clean also the other methods a bit more, there is still a lot of Code-Duplication
I did not want to overload this PR thats why I separated them

closes #1509

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+    SetBaseMembers();
+}
+
+/**
+ * Constructor using Properties
+ */
+BaseShellElement::BaseShellElement(IndexType NewId,
+    GeometryType::Pointer pGeometry,
+    PropertiesType::Pointer pProperties,
+    bool IsThickShell)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp]: Maybe use a local flag?

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 2]
```diff
@@ -0,0 +1,629 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Philipp Bucher (based on the work of Massimo Petracca and Peter Wilson)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp]: I suggest to put it in two different lines

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+        const NodeType& i_node = geom[i];
+        const array_1d<double, 3>& vel = i_node.FastGetSolutionStepValue(VELOCITY, Step);
+        // TODO also include the angular velocity
+
+        const SizeType index = i * 6;
+        rValues[index]     = vel[0];
+        rValues[index + 1] = vel[1];
+        rValues[index + 2] = vel[2];
+
+        rValues[index + 3] = 0.0;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:186]: rotation_velocity?
If you want the time scheme to work you need to define this

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+        const NodeType& i_node = geom[i];
+        const array_1d<double, 3>& acc = i_node.FastGetSolutionStepValue(ACCELERATION, Step);
+        // TODO also include the angular acceleration
+
+        const SizeType index = i * 6;
+        rValues[index]     = acc[0];
+        rValues[index + 1] = acc[1];
+        rValues[index + 2] = acc[2];
+
+        rValues[index + 3] = 0.0;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:212]: Idem for rotations

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+{
+    KRATOS_TRY;
+
+    Element::Check(rCurrentProcessInfo);
+
+    CheckVariables();
+    CheckDofs();
+    CheckProperties(rCurrentProcessInfo);
+
+    KRATOS_ERROR_IF(GetGeometry().Area() < std::numeric_limits<double>::epsilon()*1000)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:356]: It's a good practise to consider epsilon*1000?

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+///@name Protected Operators
+///@{
+
+///@}
+///@name Protected Operations
+///@{
+
+void BaseShellElement::SetBaseMembers()
+{
+    mNumDofs = 6 * GetGeometry().PointsNumber(); // 6 dofs per node
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp]: Do you really need a member variable of something that you always know and can compute in any moment?

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+        KRATOS_ERROR_IF(r_node.GetBufferSize() < 2) << "This Element needs "
+            << "at least a buffer size = 2" << std::endl;
+    }
+}
+
+void BaseShellElement::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+{
+    // check properties
+    if(pGetProperties() == nullptr)
+        KRATOS_ERROR << "Properties not provided for element " << Id() << std::endl;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:511]: KRATOS_ERROR_IF

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+
+    const PropertiesType & props = GetProperties();
+
+    const GeometryType& geom = GetGeometry(); // TODO check if this can be const
+
+    if(props.Has(SHELL_CROSS_SECTION)) // if the user specified a cross section ...
+    {
+        const ShellCrossSection::Pointer & section = props[SHELL_CROSS_SECTION];
+        if(section == nullptr)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:520]: KRATOS_ERROR_IF

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+    {
+        CheckSpecificProperties();
+
+        // perform detailed orthotropic check later in shell_cross_section
+    }
+    else // ... allow the automatic creation of a homogeneous section from a material and a thickness
+    {
+        CheckSpecificProperties();
+
+        ShellCrossSection::Pointer dummySection = ShellCrossSection::Pointer(new ShellCrossSection());
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:535]: @RiccardoRossi is not very fan of the new for shared_pointers

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+        dummySection->Check(props, geom, rCurrentProcessInfo);
+    }
+
+}
+
+void BaseShellElement::CheckSpecificProperties()
+{
+    const PropertiesType & r_props = GetProperties();
+
+    if (!r_props.Has(CONSTITUTIVE_LAW))
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:549]: KRATOS_ERROR_IF or IF_NOT

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+        KRATOS_ERROR << "CONSTITUTIVE_LAW not provided for element " << Id() << std::endl;
+    const ConstitutiveLaw::Pointer& claw = r_props[CONSTITUTIVE_LAW];
+    if (claw == nullptr)
+        KRATOS_ERROR << "CONSTITUTIVE_LAW not provided for element " << Id() << std::endl;
+
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:554]: Idem, I am not going to repeat my self

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+///@}
+///@name Serialization
+///@{
+
+void BaseShellElement::save(Serializer& rSerializer) const {
+  KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
+  rSerializer.save("Sections", mSections);
+  rSerializer.save("NumDofs", mNumDofs);
+  rSerializer.save("NumGPs", mNumGPs);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp]: Both NumGP and NimDofs can be computed in any moment

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.h [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+    /**
+    * determines the elemental list of DOFs
+    * @param ElementalDofList: the list of DOFs
+    * @param rCurrentProcessInfo: the current process info instance
+    */
+    void GetDofList(DofsVectorType& rElementalDofList, ProcessInfo& rCurrentProcessInfo) override;
+
+
+    void GetValuesVector(Vector& rValues, int Step = 0) override;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.h:120]: Missing Doxygen, idem further

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Look at my comments. Great work

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp (hunk 1 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp]: Do you have an example for this?

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp (hunk 3 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:186]: I put a todo comment to do this

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp (hunk 5 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:356]: IDK, I saw it somewhere recently
But in general I am still not decided how to use it
I also thought of using 1e-15

**Comment by philbucher** [comment]
@loumalouomega thx for the review
I agree with all of your comments, but the purpose of this PR was to unify duplicated code WITHOUT changing it.
Of course I can do this too (will do for sure in the future), but then this PR would become huge

I can do it now or later, up to @KratosMultiphysics/structural-mechanics

**Comment by loumalouomega** [comment]
Ok

**Review by loumalouomega — DISMISSED** [review]

**Comment by philbucher** [comment]
@loumalouomega I implemented some of your suggestions, the rest I will do in the next PR if this is ok for you

**Review by loumalouomega — APPROVED** [review]
Thanks

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 14]
```diff
@@ -0,0 +1,658 @@
...
+        rResult.resize(num_dofs, false);
+
+    GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        const SizeType index = i * 6;
+        NodeType& i_node = geom[i];
+
+        rResult[index]     = i_node.GetDof(DISPLACEMENT_X).EquationId();
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:101]: take a look on how to get the dofs by key (in the fluid elements for example) it is faster

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 15]
```diff
@@ -0,0 +1,658 @@
...
+    rElementalDofList.resize(0);
+    rElementalDofList.reserve(num_dofs);
+
+    GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        NodeType& i_node = geom[i];
+
+        rElementalDofList.push_back(i_node.pGetDof(DISPLACEMENT_X));
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:130]: same

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 16]
```diff
@@ -0,0 +1,658 @@
...
+    const SizeType num_dofs = GetNumberOfDofs();
+
+    if ( rDampingMatrix.size1() != num_dofs )
+        rDampingMatrix.resize( num_dofs, num_dofs, false );
+
+    noalias( rDampingMatrix ) = ZeroMatrix( num_dofs, num_dofs );
+
+    // 1.-Calculate StiffnessMatrix:
+
+    MatrixType StiffnessMatrix  = Matrix();
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:313]: you don't need to change, but note that

    MatrixType StiffnessMatrix;

would have been equivalent

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:356]: i personally prefer the 1000*eps.

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 17]
```diff
@@ -0,0 +1,658 @@
...
+    return integrationPoints.size();
+}
+
+void BaseShellElement::CalculateAll(
+    MatrixType& rLeftHandSideMatrix,
+    VectorType& rRightHandSideVector,
+    ProcessInfo& rCurrentProcessInfo,
+    const bool CalculateStiffnessMatrixFlag,
+    const bool CalculateResidualVectorFlag
+    )
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:453]: i think you need to make this virtual. Ideally virtual pure

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 18]
```diff
@@ -0,0 +1,658 @@
...
+{
+    KRATOS_ERROR << "You have called to the CalculateAll from the base class for shell elements" << std::endl;
+}
+
+ShellCrossSection::SectionBehaviorType BaseShellElement::GetSectionBehavior()
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:458]: virtual pure

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp [hunk 19]
```diff
@@ -0,0 +1,658 @@
...
+{
+    KRATOS_ERROR << "You have called to the GetSectionBehavior from the base class for shell elements" << std::endl;
+}
+
+void BaseShellElement::SetupOrientationAngles()
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:463]: can't this be implemented here?

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp (hunk 9 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp:535]: question: why do you need to make it a pointer? (if you do, please use Kratos::make_shared)

File: applications/StructuralMechanicsApplication/custom_elements/base_shell_element.h [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+protected:
+
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    IntegrationMethod mIntegrationMethod = GeometryData::GI_GAUSS_2;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_shell_element.h:200]: mmm
do you need to store this? it is one int per element, which i think is largely unneeded

File: applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp [hunk 21]
```diff
@@ -288,220 +261,53 @@ namespace Kratos
...
 		// Average mass per unit area over the whole element
 		double av_mass_per_unit_area = 0.0;
-		for (std::size_t i = 0; i < OPT_NUM_GP; i++)
+
+        const SizeType num_gps = GetNumberOfGPs();
+
+		for (SizeType i = 0; i < num_gps; i++)
 			av_mass_per_unit_area += mSections[i]->CalculateMassPerUnitArea();
-		av_mass_per_unit_area /= double(OPT_NUM_GP);
+		av_mass_per_unit_area /= double(num_gps);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp:310]: minor (and don't need to change) however if you cast, do it by 

        static_cast<double>(num_gps);

if you take a look in stack overflow there are long discussions about the why. In my opinion a convincing reason is that it makes evident where the casting is done by a simple search in the archive. As i am telling it is of minor importance though

File: applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp [hunk 22]
```diff
@@ -639,14 +423,14 @@ namespace Kratos
...
 			InitializeCalculationData(data);
 
-			// Get the current displacements in global coordinate system and 
+			// Get the current displacements in global coordinate system and
 			// transform to reference local system
 			ShellT3_LocalCoordinateSystem referenceCoordinateSystem(
 				mpCoordinateTransformation->CreateReferenceCoordinateSystem());
 			MatrixType Rdisp(18, 18);
 			referenceCoordinateSystem.ComputeTotalRotationMatrix(Rdisp);
 			data.localDisplacements = prod(Rdisp, data.globalDisplacements);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp:432]: noalias

File: applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp [hunk 23]
```diff
@@ -2492,7 +2270,7 @@ namespace Kratos
 				iValue(0, 1) = iValue(1, 0) = data.generalizedStresses[2] -
 					data.generalizedStresses[5];
 				iValue(0, 2) = iValue(2, 0) = 0.0;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp:2272]: uff, i don't understand this statement (i wouldn't be able to tell the order in which it is executed...)

File: applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.hpp [hunk 24]
```diff
@@ -365,10 +332,10 @@ namespace Kratos
 		void AddBodyForces(CalculationData& data, VectorType& rRightHandSideVector);
 
 		void CalculateAll(MatrixType& rLeftHandSideMatrix,
-			VectorType& rRightHandSideVector,
-			ProcessInfo& rCurrentProcessInfo,
-			const bool LHSrequired,
-			const bool RHSrequired);
+            VectorType& rRightHandSideVector,
+            ProcessInfo& rCurrentProcessInfo,
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.hpp:336]: btw, make the ProcessInfo const (i think this is "your" function, it is not in the base element class)

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp [hunk 25]
```diff
@@ -451,10 +268,13 @@ void ShellThinElement3D3N::CalculateMassMatrix(MatrixType& rMassMatrix, ProcessI
...
     // Calculate avarage mass per unit area
+
+    const SizeType num_gps = GetNumberOfGPs();
+
     double av_mass_per_unit_area = 0.0;
-    for(std::size_t i = 0; i < OPT_NUM_GP; i++)
+    for(std::size_t i = 0; i < num_gps; i++)
         av_mass_per_unit_area += mSections[i]->CalculateMassPerUnitArea();
-    av_mass_per_unit_area /= double(OPT_NUM_GP);
+    av_mass_per_unit_area /= double(num_gps);
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp:277]: static_cast (minor)

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp [hunk 26]
```diff
@@ -1261,11 +1051,13 @@ void ShellThinElement3D3N::InitializeCalculationData(CalculationData& data)
     // that already take into accout the
     // thickness
 
-    data.dA = A / (double)OPT_NUM_GP;
+    const SizeType num_gps = GetNumberOfGPs();
+
+    data.dA = A / (double)num_gps;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp:1056]: static_cast

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp [hunk 27]
```diff
@@ -2059,7 +1854,7 @@ bool ShellThinElement3D3N::TryGetValueOnIntegrationPoints_GeneralizedStrainsOrSt
 			iValue(1, 1) = data.generalizedStresses(1) +
 				data.generalizedStresses(4);
 			iValue(2, 2) = 0.0;
-			iValue(0, 1) = iValue(1, 0) = data.generalizedStresses[2] + 
+			iValue(0, 1) = iValue(1, 0) = data.generalizedStresses[2] +
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.cpp:1857]: unclear to me...

File: applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.hpp [hunk 28]
```diff
@@ -67,8 +66,6 @@ class ShellThinElement3D3N : public Element
 
     KRATOS_CLASS_POINTER_DEFINITION(ShellThinElement3D3N);
 
-    typedef std::vector< ShellCrossSection::Pointer > CrossSectionContainerType;
-
     typedef ShellT3_CoordinateTransformation CoordinateTransformationBaseType;
 
     typedef Kratos::shared_ptr<CoordinateTransformationBaseType> CoordinateTransformationBasePointerType;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/shell_thin_element_3D3N.hpp:71]: typedef typename CrossSectionContainerType::Pointer CoordinateTransformationBasePointerType;

**Review by RiccardoRossi — COMMENTED** [review]
some minor remarks in my review. please take a look before merging.

one major one: what about the LOCAL_AXES? i mean, the MATERIAL_ANGLE should be given wrt such local axes no?

**Comment by RiccardoRossi** [comment]
(btw, thx for the time it took doing this... i think it is a very nice improvement!!)

**Comment by philbucher** [comment]
@RiccardoRossi thx for the review.
As I said already I was planning to do the cleanup in several steps but since you guys seem eager to review I can do some more in this PR ... there is a lot to do...
This will take a few days though

**Comment by loumalouomega** [comment]
@philbucher are we waiting for any reason?

**Comment by philbucher** [comment]
Well I will add a lot more to this branch, but this takes time ...

**Comment by RiccardoRossi** [comment]
hi Philip,

if this is operative, maybe you'd best merge what is there, and later on add more (as a different PR).

I guess this was your original idea and sounds good to me.

**Comment by RiccardoRossi** [comment]
hi Philip,

if this is operative, maybe you'd best merge what is there, and later on add more (as a different PR).

I guess this was your original idea and sounds good to me.

**Comment by philbucher** [comment]
it is of course operative
I will merge it now and take your comments into consideration in the next iteration.

**Pull request merged**